t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+            sem = nil
+            local semaphore = require "ngx.semaphore"
+            ngx.say(g.err)
+            collectgarbage("collect")
+       ';
+    }
+--- request
+GET /test
+--- response_body
+success
+--- log_level: debug
+--- error_log
+ngx_http_lua_ffi_semaphore_gc
+
+
+
+=== TEST 14: semaphore post in body_filter_by_lua
+--- http_config eval: $::HttpConfig
+--- config
+    location /test{
+        content_by_lua '
+            local res1,res2 = ngx.location.capture_multi{
+                {"/sem_wait"},
+                {"/sem_post"},
+            }
+            ngx.say(res1.status)
+            ngx.say(res1.body)
+            ngx.say(res2.status)
+            ngx.say(res2.body)
+        ';
+    }
+
+    location /sem_wait {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:wait(10)
+            if ok then
+                ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.exit(500)
+            end
+        ';
+    }
+
+    location /sem_post {
+        body_filter_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.log(ngx.ERR, err)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:post()
+            if not ok then
+                ngx.log(ngx.ERR, err)
+            end
+    ';
+    content_by_lua '
+            ngx.print("post")
+            ngx.exit(200)
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- log_level: debug
+--- no_error_log
+[error]
+
+
+
+=== TEST 15: semaphore post in log_by_lua
+--- http_config eval: $::HttpConfig
+--- config
+        location /test {
+        log_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.log(ngx.ERR, err)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:post()
+            if not ok then
+                ngx.log(ngx.ERR, err)
+            end
+    ';
+    content_by_lua '
+            ngx.say("post")
+            ngx.exit(200)
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+post
+--- no_error_log
+[error]
```

Comment:
The checks in this test case is too weak. This case can pass even when the whole `log_by_lua` thing is not run at all. Let's add some outputs along the OK code paths as well and check them.